### DISCUSSION TITLE
Do I need to configure renovate in a particular way to support private ghcr docker repositories?

### DISCUSSION THREAD

**Discussion by gabrielsadaka** [renovatebot/renovate#25964] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

Do I need to configure renovate in a particular way to support private ghcr docker repositories?

I have a `docker-compose.yaml` that Renovate has correctly detected and determined that the docker images are stored in a `ghcr.io` registry. 

Both the repository and registry are private within a private Github organisation. 

I have provided the log output from Renovate that states `Could not determine new digest for update`. 

Here is my `renovate.json` config:

```
{
  "extends": [
    "config:base"
  ],
  "prHourlyLimit": 0,
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "automerge": true
    }
  ]
}
```

### Logs (if relevant)

<details><summary>Logs</summary>

```
          {
            "depName": "ghcr.io/org/postgres",
            "currentDigest": "sha256:32f788f238c74567ebe8c42d1b145501165f26ea4333b4b9f225cd58e7df3780",
            "replaceString": "ghcr.io/org/postgres@sha256:32f788f238c74567ebe8c42d1b145501165f26ea4333b4b9f225cd58e7df3780",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "updates": [],
            "packageName": "ghcr.io/org/postgres",
            "versioning": "docker",
            "warnings": [
              {
                "message": "Could not determine new digest for update (docker package ghcr.io/org/postgres)",
                "topic": "ghcr.io/org/postgres"
              }
            ]
          },

```

</details>

**Comment by rarkins** [comment, marked as answer]
Yes, you will need to configure hostRules with matchHost=ghcr.io with a personal access token which has appropriate access. For now the app token which Renovate has does not work for ghcr.io

**Reply by gabrielsadaka** [reply]
Thanks, I tried adding the following config but I still get the same error:

```
  "hostRules": [
    {
      "matchHost": "ghcr.io",
      "username": "gabrielsadaka",
      "hostType": "docker",
      "encrypted": {
        "password": "ENCRYPTED_PASSWORD by https://app.renovatebot.com/encrypt"
      }
    }
  ]
  ```
  
  Is that correct?

**Reply by gabrielsadaka** [reply]
@rarkins I still haven't been able to get this to work. I have tested the personal access token directly and it works fine. Not sure why Renovate is receiving a 404 from the manifests endpoint. Is there anything I have done wrong?

```
"options": {
      "headers": {
        "user-agent": "RenovateBot/37.220.2 (https://github.com/renovatebot/renovate)",
        "authorization": "***********",
        "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://ghcr.io/v2/org/postgres/manifests/latest",
      "hostType": "docker",
      "username": "",
      "password": "",
      "method": "HEAD",
      "http2": false
    },
    "response": {
      "statusCode": 404,
      "statusMessage": "Not Found",
      "body": "",
      "headers": {
        "content-type": "application/json",
        "docker-distribution-api-version": "registry/2.0",
        "date": "Tue, 05 Mar 2024 06:35:02 GMT",
        "content-length": "70",
        "x-github-request-id": "C473:6E9A:9B964:DB261:65E6BD16",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
  "registryHost": "https://ghcr.io"
  "dockerRepository": "org/postgres"
  "tag": "latest"
}
```

**Reply by gabrielsadaka** [reply]
After getting the auth working, the next issue was that the images being pushed to ghcr were missing the `latest` tag, hence the 404 Renovate was logging. 

Adding the latest tag to the metadata github action that is part of the job that builds and pushes the images fixed it. 

```
      - name: Extract metadata (tags, labels) for Docker
        id: meta
        uses: docker/metadata-action@232fc64e3a4e54539e087c5976439ea54be0959d
        with:
          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
          tags: |
            # set latest tag for default branch
            type=raw,value=latest,enable={{is_default_branch}}
            type=sha
```

**Reply by lapwat** [reply]
@gabrielsadaka 
Can you explain how this works please?

This fixed my renovate workflow. But this is only **extracting** metadata from a Container Registry. How does that **set** _latest tag for default branch_ ?

**Reply by gabrielsadaka** [reply]
> @gabrielsadaka 
> Can you explain how this works please?
> 
> This fixed my renovate workflow. But this is only **extracting** metadata from a Container Registry. How does that **set** _latest tag for default branch_ ?

It's poorly named, the documentation for the metadata GitHub action states: "GitHub Action to extract metadata from Git reference and GitHub events. This action is particularly useful if used with Docker Build Push action to tag and label Docker images."

So it's not extracting the metadata from the registry. It's intended to extract it from the git reference and GitHub event metadata, then apply that to the image that was built, so those tags are available when it is pushed to the registry.

**Comment by dargmuesli** [comment]
I found the following setup in the UI to be working:

<img width="587" height="564" alt="image" src="https://github.com/user-attachments/assets/e4354fec-1a53-45b7-89f0-0fb00e91fe7b" />
